add config flag and update readme

gabrielthomasjacobs opened the pull request:
As discussed, added support for a config file.

The planned layout of the config will be:

```
{
"mvp": {...mocha-vite-puppeteer-opts},
"mocha": {...mocha-opts},
"vite": {...vite-opts},
"puppeteer": {...puppetteer-opts}
}
```
Where "mvp" are any of the currently accepted cmd flags

the json file can be added to the script/npx call by using the --config flag or "-c" followed by the relative path to the json file. Should behave the same as the existing "entry" input.

As of this PR, only the keys "mvp" and "puppeteer" are supported. Any options defined in the config will override the options specified in the command line flags (with the exception of debug overriding the headless option to false). I figured we would add the other keys as we need to support them.

Going forward, this should be a good place to pass-through any advanced configs, with the bonus of looking cleaner when using many options.

gabrielthomasjacobs commented:
Also, please let me know if you'd like me to "punch-out" the options handling into it's own js file. Might keep the main cli cleaner.

larsthorup commented:
Thank you for the PR! I have a few questions / comments:

- Please fix the build :)
- I am not too happy with the "mvp" key name. The name of this tool should be consistent. Either we can use "mocha-vite-puppeteer", which you might find too long. Or we can decide to keep the direct tool options at the top-level (not nested), which I think could work well.
- I agree with the gradual adoption of options for the embedded tools, while at the same time having a clear plan for how to extend this adoption as the needs arise.
- Could you adapt the test to ensure that the config parsing is tested during build?

larsthorup commented:
Also, please merge from "main", as I have updated our dependencies.

gabrielthomasjacobs commented:
Ok, build is fixed (I was doing a ternary check on a string 'undefined'... 🙄)

I've removed the mvp key and just set the options to the base-level of the config object. I was initially concerned about overlapping keys, but since we're placing the other settings under 'puppeteer' or 'mocha', that won't be an issue.  

I've updated the readme to reflect this change.

I've also duplicated the test-flag test setup and just modified it to test the config file. I've updated the npm script "test" to include this.

And this branch is rebased from master!

gabrielthomasjacobs commented:
Oh, and I've changed the _options_ variable from a const to a let. After it is read in from the flags/cmd, it's overwritten by the json file. From there, we can just use _options_ as the main variable that'll store all the user options.

larsthorup commented on cli.js:
```diff
@@ -10,7 +10,7 @@ import { createServer } from 'vite'
 import { MochaProtocolReporter } from './mochaProtocolReporter.js';
 
-const options = getopt({
+let options = getopt({
```
I think it can stay "const", since you are not reassigning the options variable itself, only the keys. `const` only applies to top level.

larsthorup commented on cli.js:
```diff
@@ -24,10 +24,10 @@ if(options.args) {
+const jsonOptions = options.config && options.config !== 'undefined' ? JSON.parse(fs.readFileSync(options.config, 'utf-8')) : undefined;
+if(jsonOptions) {
+  Object.keys(jsonOptions).forEach(key => {
+    options[key] = jsonOptions[key];
```
I didn't notice this previously, but I think the normal way is for CLI options to take precedence over config file options, where this code makes it the other way around. If you agree, would you mind changing it?

gabrielthomasjacobs replied:
Ok, I've switched the order

I've removed the default key from the stdio getopts method, since that would specify a value and we'd use that over anything specified in the json. 

If we'd prefer to keep a default specified and then do some logic to prefer the json whenever the cli value matches the default value, that might be nicer. 
If we take that approach though, if a user explicitly specified a default value, it would be ignored in favor of a json value with the same key. E.g. if you set the port as 3006 in json and then try to manually set it to 3001 in the flags, 3001 would be ignored (since 3001 is the default).

larsthorup replied:
I like this implementation. Now the precedence is: CLI options OVER config file options OVER default values. Thx!

gabrielthomasjacobs commented:
Also, I plan on refactoring the options handling to a new file. I'll do that in a new PR once this is merged. Do you have a preference for file name? I was thinking just `optionsParser.js`
